src/vs/workbench/contrib/chat/browser/chatContentParts/toolInvocationParts/chatTerminalToolProgressPart.ts

```diff
@@ -754,168 +927,321 @@ class ChatTerminalToolOutputSection extends Disposable {
+		this._resetLineState();
+		this._lastOutputTruncated = false;
+		storedOutput.text = '';
+		storedOutput.truncated = false;
+
+		const hasDetached = !!(this._detachedTerminal.value && this.isExpanded);
+		if (hasDetached) {
+			this._clearDetachedTerminal();
+			this._needsReplay = false;
+		} else {
+			this._needsReplay = true;
+		}
+
+		if (!snapshot) {
+			this._setStatusMessages();
+			this._updateTerminalVisibility();
 			return;
 		}
-		const result = await xterm.getCommandOutputAsHtml(command, CHAT_TERMINAL_OUTPUT_MAX_PREVIEW_LINES);
-		return { text: result.text, truncated: result.truncated ?? false };
+
+		this.appendStreamingData(snapshot);
 	}
 
-	private _getStoredCommandOutput(): { text: string; truncated: boolean } | undefined {
-		const stored = this._terminalData.terminalCommandOutput;
-		if (!stored?.text) {
+	private _clearDetachedTerminal(): void {
+		// Clears the detached xterm prior to replaying content so the terminal reflects the latest
+		// snapshot exactly.
+		const instance = this._detachedTerminal.value;
+		if (!instance) {
 			return;
 		}
-		return {
-			text: stored.text,
-			truncated: stored.truncated ?? false
-		};
+		const xterm = instance.xterm as unknown as XtermTerminal | undefined;
+		if (!xterm) {
+			return;
+		}
+		try {
+			xterm.raw.clear();
+			xterm.write('\x1b[3J\x1b[2J\x1b[H');
+		} catch {
+			// The detached terminal may be mid-dispose; ignore errors when clearing.
+		}
+	}
+
+	public beginStreaming(): void {
+		// Resets streaming state just before a command starts emitting fresh data.
+		this._isStreaming = true;
+		this._streamBuffer = [];
+		this._resetLineState();
+		this._lastOutputTruncated = false;
+		this._lastRawSnapshot = undefined;
+		this._needsReplay = true;
+		if (this._detachedTerminal.value) {
+			this._clearDetachedTerminal();
+		}
+		this._terminalData.terminalCommandOutput = { text: '', truncated: false };
+		this._setSupplementalMessages([]);
+		this._scrollable.scanDomNode();
+		this._updateTerminalVisibility();
+	}
+
+	public endStreaming(): void {
+		this._isStreaming = false;
+		this._setStatusMessages();
+		this._updateTerminalVisibility();
+	}
+
+	public hasRenderableOutput(): boolean {
+		return this._hasRenderableOutput();
+	}
+
+	public get isOutputTruncated(): boolean {
+		return this._lastOutputTruncated;
 	}
 
-	private _renderOutput(result: { text: string; truncated: boolean }): { element: HTMLElement; inlineOutput?: HTMLElement; pre?: HTMLElement } {
-		this._lastOutputTruncated = result.truncated;
-		const { content } = h('div.chat-terminal-output-content@content');
-		let inlineOutput: HTMLElement | undefined;
-		let preElement: HTMLElement | undefined;
+	private _hasRenderableOutput(): boolean {
+		return this._streamBuffer.some(chunk => {
+			const withoutCsi = chunk.replace(CSI_SEQUENCE_REGEX, '');
+			const withoutAnsi = removeAnsiEscapeCodes(withoutCsi);
+			return withoutAnsi.replace(/\r/g, '').trim().length > 0;
+		});
+	}
```

Comment:
`removeAnsiEscapeCodes` should remove CSI too:

https://github.com/microsoft/vscode/blob/7e98d11d64b4d8d1c6aa60d72c5c6c27ecced23c/src/vs/base/common/strings.ts#L788-L796

We should use events for this instead of doing a bunch of string manipulations to try determine it. If `onWriteParsed`/`onCursorMove`/`onRender` from xterm don't give us what we need we could expose another event.